src/Stream/GetProductsStreamer.php

```diff
@@ -0,0 +1,295 @@
+                $metadata[self::METADATA_IS_LAST_PAGE] = $isLastPage;
+            },
+            '/' . self::METADATA_LAST_PAGE_NUMBER => function ($lastPageNumber) use (&$metadata) {
+                $metadata[self::METADATA_LAST_PAGE_NUMBER] = $lastPageNumber;
+            },
+            '/' . self::METADATA_FIRST_ON_PAGE => function ($firstOnPage) use (&$metadata) {
+                $metadata[self::METADATA_FIRST_ON_PAGE] = $firstOnPage;
+            },
+            '/' . self::METADATA_LAST_ON_PAGE => function ($lastOnPage) use (&$metadata) {
+                $metadata[self::METADATA_LAST_ON_PAGE] = $lastOnPage;
+            },
+            '/' . self::METADATA_ITEMS_PER_PAGE => function ($itemsPerPage) use (&$metadata) {
+                $metadata[self::METADATA_ITEMS_PER_PAGE] = $itemsPerPage;
+            }
+        ]);
+
+        $parser = new Parser($stream, $listener);
+        $parser->parse();
+
+        $this->closeStream($stream);
+
+        return $metadata;
+    }
+
+    /**
+     * @param string $streamedFileName
+     * @return int
+     */
+    public function getProductsCount($streamedFileName) {
+        $metadata = $this->getMetadata($streamedFileName);
+
+        if (!$metadata[self::METADATA_IS_LAST_PAGE]) {
+            return $metadata[self::METADATA_ITEMS_PER_PAGE];
+        }
+
+        return abs($metadata[self::METADATA_ITEMS_PER_PAGE] * $metadata[self::METADATA_PAGE] - $metadata[self::METADATA_TOTALS_COUNT]);
+    }
+
+    /**
+     * @param $streamedFileName
+     * @return array|null
+     */
+    public function getProductsIds($streamedFileName)
+    {
+        $callbackResult = [];
+
+        $stream = $this->openStream($streamedFileName);
+
+        $productIdListener = new RegexListener([
+            '/entries/*/id' => function ($productId) use (&$callbackResult) {
+                $callbackResult[] = $productId;
+            }
+        ]);
+
+        $parser = new Parser($stream, $productIdListener);
+        $parser->parse();
+
+        $this->closeStream($stream);
+
+        return $callbackResult;
+    }
+
+    /**
+     * @param $streamedFileName
+     * @param array $validItemIds
+     * @param array $variations
+     * @param $createProduct
+     * @return array
+     */
+    public function processProducts($streamedFileName, $validItemIds, $variations, callable $createProduct)
+    {
+        $productsCount = $this->getProductsCount($streamedFileName);
+
+        if ($productsCount <= 0) {
+            return [];
+        }
+
+        $trackSkippedProductsIds = [];
+
+        for ($i = 0; $i < $productsCount; $i++) {
+            $stream = $this->openStream($streamedFileName);
+
+            $getProductListener = new RegexListener();
+            $parser = new Parser($stream, $getProductListener);
+            $getProductListener->setMatch([
+                '/entries/' . $i  => function ($productData) use ($createProduct, $variations, $validItemIds, &$trackSkippedProductsIds, $parser) {
```

Comment:
Regex is not precise enough since it gets the whole product data.
We may need to refactor how a product gets created, see `createProductItem()`.
The stream should be passed in the constructor of `Product.php` and the getter should take care to read the correct data with regexlistener out of the stream.

Reply:
The stream should not parse a whole product and write each values to the csv, we should already get each piece of necessary product data out of the stream with more precise regex and put it to the csv